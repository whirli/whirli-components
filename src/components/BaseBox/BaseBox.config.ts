// @ts-ignore
import { PropType } from '@composition';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseBox/BaseBox.config';

// Package config
import { PropKeys } from './BaseBox.constants';
import type { PropTag, PropPadding } from './BaseBox.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  [PropKeys.TAG]: {
    required: false,
    type: String as PropType<PropTag>,
    default: 'div',
  },
  [PropKeys.PADDING]: {
    required: false,
    type: [String, Array] as PropType<PropPadding | PropPadding[]>,
    default: 'base',
  },
  ...config.props,
};
