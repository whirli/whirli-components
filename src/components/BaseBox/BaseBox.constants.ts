export const PropKeys: Record<string, string> = {
  TAG: 'tag',
  PADDING: 'padding',
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.TAG]: ['div'],
  [PropKeys.PADDING]: ['none', '2xs', 'xs', 'sm', 'base', 'md', 'lg', 'xl', '2xl'],
};
