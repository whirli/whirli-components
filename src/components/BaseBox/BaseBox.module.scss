@import '@whirli-components/styles/Breakpoints.module.scss';
@import '@whirli-components/styles/Spacers.scss';

// default styles
.base-box {
    width: 100%;
}

$sides: top, right, bottom, left;

@each $side in $sides {
    .padding-#{$side}-none {
        padding-#{$side}: $spacerNone;
    }
    .padding-#{$side}-2xs {
        padding-#{$side}: $spacer2Xs;
    }
    .padding-#{$side}-xs {
        padding-#{$side}: $spacerXs;
    }
    .padding-#{$side}-sm {
        padding-#{$side}: $spacerSm;
    }
    .padding-#{$side}-base {
        padding-#{$side}: $spacerDefault;
    }
    .padding-#{$side}-md {
        padding-#{$side}: $spacerMd;
    }
    .padding-#{side}-lg {
        padding-#{$side}: $spacerLg;
    }
    .padding-#{$side}-xl {
        padding-#{$side}: $spacerXl;
    }
    .padding-#{$side}-2xl {
        padding-#{$side}: $spacer2Xl;
    }
}

// breakpoint styles
$packageBreakpointStyles: (
    /*color-black: (
        color: black,
    ),*/
);