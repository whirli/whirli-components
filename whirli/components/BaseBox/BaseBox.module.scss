@import '@whirli-components/components/BaseBox/BaseBox.module.scss';

// Standard styles

// Breakpoint styles
$localBreakpointStyles: ();

@include createBreakpointClasses(map-merge($packageBreakpointStyles, $localBreakpointStyles));