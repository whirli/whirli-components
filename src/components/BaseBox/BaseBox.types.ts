import { PropKeys, PropValues } from './BaseBox.constants';

const propTag: string[] = PropValues[PropKeys.TAG];
const propPadding: string[] = PropValues[PropKeys.propPadding];

export type PropTag = typeof propTag[number];
export type PropPadding = typeof propPadding[number];
