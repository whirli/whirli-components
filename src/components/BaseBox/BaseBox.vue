<template>
  <component :is="props.tag" :class="classes"><slot /></component>
</template>

<script setup lang="ts">
// Vue
import { computed, ComputedRef } from '@composition';
// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseBox/BaseBox.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseBox.config';

// Types
import { ComponentStyles as ComponentStylesInterface } from '@whirli-components/@types/components';
import { Props } from '@whirli-components/@types/props';

// Composables
import useClasses from '@whirli-components/@use/class';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

// @ts-ignore
const props: Props = defineProps(ConfigProps);

// Classes
const { makeClasses } = useClasses();

const padding = (size: string, side: string) => styles[`padding-${side}-${size}`];

const paddingClasses: ComputedRef<string[]> = computed(() => {
  if (props.padding.length === 1 || typeof props.padding === 'string') {
    return [
      padding(props.padding, 'top'),
      padding(props.padding, 'right'),
      padding(props.padding, 'bottom'),
      padding(props.padding, 'left'),
    ];
  }
  if (padding.length === 2) {
    return [
      padding(props.padding[0], 'top'),
      padding(props.padding[1], 'right'),
      padding(props.padding[0], 'bottom'),
      padding(props.padding[1], 'left'),
    ];
  }
  if (padding.length === 3) {
    return [
      padding(props.padding[0], 'top'),
      padding(props.padding[1], 'right'),
      padding(props.padding[2], 'bottom'),
      padding(props.padding[1], 'left'),
    ];
  }
  if (padding.length === 4) {
    return [
      padding(props.padding[0], 'top'),
      padding(props.padding[1], 'right'),
      padding(props.padding[2], 'bottom'),
      padding(props.padding[3], 'left'),
    ];
  }
  return [];
});

const classes: ComputedRef<string[]> = computed(() => [
  styles['base-box'],
  ...paddingClasses.value,
  makeClasses(ComponentStyles, props, styles),
]);
</script>
