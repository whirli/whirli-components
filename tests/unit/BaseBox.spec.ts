import { shallowMount } from '@vue/test-utils';

import BaseBox from '@/components/BaseBox/BaseBox.vue';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let wrapper: any;
const text = 'Hello world';

describe('BaseBox.vue', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseBox, {
      props: {
        padding: 'md',
      },
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the correct content', () => {
    expect(wrapper.find('div').text()).toBe(text);
  });
});
