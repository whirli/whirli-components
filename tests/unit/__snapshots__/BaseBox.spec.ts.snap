// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseBox.vue renders the component 1`] = `
<div
  class=""
>
  
  Hello world
  
</div>
`;
